persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cassandra/CassandraReadSideHandler.scala

```diff
@@ -0,0 +1,172 @@
+    handler.asInstanceOf[(Event, Offset) => CompletionStage[JList[BoundStatement]]].apply(event, offset).toScala.map { statements =>
+      TreePVector.from(statements).plus(offsetStore.writeOffset(event.aggregateTag.tag, offset)).asInstanceOf[JList[BoundStatement]]
+    }.toJava
+  }
+
+  override def globalPrepare(): CompletionStage[Done] = {
+    Future.successful(globalPrepareCallback).flatMap(_.apply().toScala).flatMap { _ =>
+      offsetStore.globalPrepare
+    }.toJava
+  }
+
+  override def prepare(tag: AggregateEventTag[Event]): CompletionStage[Offset] = {
+    Future.successful(prepareCallback).flatMap(_.apply(tag).toScala).flatMap { _ =>
+      offsetStore.prepare(tag.tag)
+    }.toJava
+  }
+}
+
+private[cassandra] class LegacyCassandraReadSideHandler[Event <: AggregateEvent[Event]](
+  session:            CassandraSession,
+  cassandraProcessor: CassandraReadSideProcessor[Event],
+  dispatcher:         String
+)(implicit ec: ExecutionContext) extends CassandraReadSideHandler[Event, BiFunction[_ <: Event, UUID, CompletionStage[JList[BoundStatement]]]](
+  session,
+  cassandraProcessor.defineEventHandlers(new cassandraProcessor.EventHandlersBuilder).handlers,
+  dispatcher
+) {
+
+  override protected def invoke(handler: BiFunction[_ <: Event, UUID, CompletionStage[JList[BoundStatement]]], event: Event, offset: Offset): CompletionStage[JList[BoundStatement]] = {
+    offset match {
+      case uuid: TimeBasedUUID => handler.asInstanceOf[BiFunction[Event, UUID, CompletionStage[JList[BoundStatement]]]].apply(event, uuid.value)
+      case other               => throw new IllegalArgumentException("CassandraReadSideProcessor does not support offsets of type " + other.getClass.getName)
+    }
+  }
+
+  override def prepare(tag: AggregateEventTag[Event]): CompletionStage[Offset] = {
+    Future.successful(cassandraProcessor).flatMap(_.prepare(session).toScala).map { maybeUuid =>
+      if (maybeUuid.isPresent) {
+        Offset.timeBasedUUID(maybeUuid.get)
+      } else Offset.NONE
+    }.toJava
+  }
+}
+
+private[cassandra] class OffsetStore(session: CassandraSession, offsetTable: String)(implicit ec: ExecutionContext) {
+  private var writeOffsetStatement: PreparedStatement = _
+
+  def globalPrepare = {
+    session.executeCreateTable(s"""
+                                  |CREATE TABLE IF NOT EXISTS $offsetTable (
+                                  |  partition text, timeUuidOffset timeuuid, sequenceOffset bigint,
+                                  |  PRIMARY KEY (partition)
+                                  |)""".stripMargin).toScala
+  }
+
+  def prepare(partition: String) = {
+    prepareWriteOffset
+      .flatMap(_ => readOffset(partition))
+  }
+
+  private def prepareWriteOffset: Future[Done] = {
+    session.prepare(s"INSERT INTO $offsetTable (partition, timeUuidOffset, sequenceOffset) VALUES (?, ?, ?)").toScala.map { ps =>
+      writeOffsetStatement = ps
```

Comment:
Not on the one offset handler, but there may be multiple offset handlers (one for each shard). But to address your concerns, no, this method will only be executed once per instance of this object.